### PR TITLE
sc-4803 Initial rVASP server tests

### PR DESCRIPTION
**Pull request by pdeziel**
This adds some testing infrastructure to be able to make rVASP `Transfer` RPC calls and verify that the rVASP is returning expected responses.

### CODE REVIEW

File: pkg/rvasp/bufconn/listener.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	return g.Listener.Dial()
+}
+
+func (g *GRPCListener) Close() {
+	g.Conn.Close()
+}
+
+func (s *GRPCListener) Release() {
+	s.Listener.Close()
+}
```
Comment (pdeziel) [inline pkg/rvasp/bufconn/listener.go:44]: Copied from the other code bases but using the "alice" name to get the `FromContext` to work.

File: pkg/rvasp/transfer.go [hunk 2]
```diff
@@ -159,6 +159,12 @@ func parsePayload(payload *protocol.Payload, response bool) (identity *ivms101.I
 		return nil, nil, nil, fmt.Errorf("could non unmarshal identity: %s", err)
 	}
 
+	// Validate identity fields
+	if identity.Originator == nil || identity.OriginatingVasp == nil || identity.BeneficiaryVasp == nil || identity.Beneficiary == nil {
+		log.Warn().Msg("incomplete identity payload")
+		return nil, nil, nil, fmt.Errorf("incomplete identity payload")
+	}
```
Comment (bbengfort) [inline pkg/rvasp/transfer.go:166]: How does this validation work with the repair policies?

File: pkg/rvasp/mock.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	protocol.RegisterTRISANetworkServer(s.srv, s)
+
+	return s, remotePeers, mockDB, s.certs, s.chain, nil
+}
+
+// NewServerMock returns a mock rVASP server that can be used for testing.
+func NewServerMock(conf *config.Config) (s *Server, mockDB sqlmock.Sqlmock, err error) {
+	s = &Server{conf: conf, echan: make(chan error, 1)}
+	if s.db, mockDB, err = db.NewDBMock("alice"); err != nil {
+		return nil, nil, err
```
[the remark this answers is not in the document]
Comment (bbengfort) [reply pkg/rvasp/mock.go:55]: That is a bit awkward; one thing I've been doing is type assertions for mocks so that I don't have to track these pieces every where and I can just get access to them. For example;

```go
// RealDB implements the DB interface
type RealDB struct {
    conn sql.DB
}

// MockDB implements the DB interface
type MockDB struct {
    mock mock.DB   
}


mockDB := db.(*MockDB).GetMock()
```

This isn't quite as easy as it looks, but if done right, does keep the accessors in the right place and increases the encapsulation. 

Also, I like the [mocking and testing article](https://medium.com/safetycultureengineering/flexible-mocking-for-testing-in-go-f952869e34f5) that I posted in #blogfodder a while ago that you return the type not the interface, keeping interfaces small and asserting that a type meets a certain interface. This pattern can emerge from what I'm showing above, but like I said, it does take a bit of effort.

File: pkg/rvasp/mock.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		return nil, nil, nil, nil, nil, err
+	}
+	s.srv = grpc.NewServer(creds)
+	protocol.RegisterTRISANetworkServer(s.srv, s)
+
+	return s, remotePeers, mockDB, s.certs, s.chain, nil
+}
+
+// NewServerMock returns a mock rVASP server that can be used for testing.
+func NewServerMock(conf *config.Config) (s *Server, mockDB sqlmock.Sqlmock, err error) {
```
Comment (bbengfort) [inline pkg/rvasp/mock.go:52]: Is this just a stub right now and currently not used for tests?

**Review by bbengfort — APPROVED** [review]
I didn't get a chance to dive into the specific tests since I have to run to take my son to fencing lessons but I didn't want to leave without a :shipit:  -- I reviewed the testing framework, mocks and set up and feel pretty comfortable with what you have. It does seem like there is a lot of set up for each test, but that's potentially unavoidable. 

Thanks for all the hard work putting these tests together!

File: pkg/rvasp/transfer.go (hunk 2 above)
Comment (pdeziel) [reply pkg/rvasp/transfer.go:166]: Hmm, my assumption was that if any of these fields are nil then the identity payload is invalid. Are you saying we should be more forgiving on the repair policy and allow payloads which have nil fields?

File: pkg/rvasp/mock.go (hunk 4 above)
Comment (pdeziel) [reply pkg/rvasp/mock.go:52]: It's not used directly by the tests, but `NewTRISAMock` depends on it. I figured it's a logical division of labor since it kind of mirrors what the real code is doing, although this would have to be modified (to start the bufconn server and such) a bit if we want to write tests to verify that the originator side is sending valid envelopes.

File: pkg/rvasp/mock.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pdeziel) [reply pkg/rvasp/mock.go:55]: Interesting, I'll keep this in mind.

File: pkg/rvasp/transfer.go (hunk 2 above)
Comment (bbengfort) [reply pkg/rvasp/transfer.go:166]: If the incoming payload is partial (Beneficiary and Beneficiary VASP are nil) and the policy is repair, then the rVASP should populate the beneficiary information into the payload. 

Also - I'm not 100% sure that nil is a good check for deserialized protocol buffers; I've been caught out on weird edge cases where I receive the zero-valued struct rather than nil for (I believe) nested messages. It's probably something we'd want to test specifically.

File: pkg/rvasp/mock.go (hunk 4 above)
Comment (bbengfort) [reply pkg/rvasp/mock.go:52]: Makes sense, thanks!
